gcc/rust/hir/rust-ast-lower-pattern.cc

```diff
@@ -259,5 +259,24 @@ ASTLoweringPattern::visit (AST::GroupedPattern &pattern)
 			       pattern.get_locus ());
 }
 
+void
+ASTLoweringPattern::visit (AST::ReferencePattern &pattern)
+{
+  auto crate_num = mappings->get_current_crate ();
+  Analysis::NodeMapping mapping (crate_num, pattern.get_node_id (),
+				 mappings->get_next_hir_id (crate_num),
+				 UNKNOWN_LOCAL_DEFID);
+
+  HIR::Pattern *inner
+    = ASTLoweringPattern::translate (pattern.get_referenced_pattern ().get ());
+
+  translated
+    = new HIR::ReferencePattern (mapping, std::unique_ptr<HIR::Pattern> (inner),
+				 pattern.get_is_mut () ? Mutability::Mut
+						       : Mutability::Imm,
+				 pattern.get_has_two_amps (),
+				 pattern.get_locus ());
```

Comment:
Not for this PR, but we could probably think about desugaring double references here into an `HIR::ReferencePattern(mappings, HIR::ReferencePattern(...), mut, ...)` and remove the `has_two_amps` field of the `HIR::ReferencePattern` class. In the AST, having the flag is useful and helpful because the lexer does not know what a double reference is, so you end up going through either an `AMP` ('&') token or a... `LOGICAL_AND` ('&&') token :) but I think this is a good point to strip away that obligation since the HIR does not need to know about this.

Reply:
Let me know if you're interested in working on this @powerboat9